x/twap/api_test.go

```diff
@@ -113,100 +160,201 @@ func (s *TestSuite) TestGetArithmeticTwap() {
+		},
+		"(3 pairs of 3 records, three asset pool) start at second record, end after third record, sp1": {
+			recordsToSet: []types.TwapRecord{
+				tapRecordAB, tapRecordAC, tapRecordBC,
+				tPlus10sp5TapRecordAB, tPlus10sp5TapRecordAC, tPlus10sp5TapRecordBC,
+				tPlus20sp2TapRecordAB, tPlus20sp2TapRecordAC, tPlus20sp2TapRecordBC},
+			ctxTime: tPlusOneMin,
+			input:   makeSimpleTapTwapInput(baseTime.Add(10*time.Second), baseTime.Add(30*time.Second), quoteAssetB),
+			expTwap: []sdk.Dec{sdk.NewDecWithPrec(35, 2), sdk.NewDecWithPrec(35, 2), sdk.NewDecWithPrec(35, 2)}, // 0.2 for 10s, 0.5 for 10s
 		},
 		// start in middle of first and second record, end in middle of second and third record
-		"(3 record) interpolate: in between second and third record": {
+		"(3 records) interpolate: in between second and third record": {
 			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record, tPlus20sp2Record},
 			ctxTime:      tPlusOneMin,
 			input:        makeSimpleTwapInput(baseTime.Add(15*time.Second), baseTime.Add(25*time.Second), quoteAssetA),
-			expTwap:      sdk.NewDecWithPrec(35, 1), // 5 for 5s, 2 for 5 = 35 / 10 = 3.5
+			expTwap:      []sdk.Dec{sdk.NewDecWithPrec(35, 1)}, // 5 for 5s, 2 for 5 = 35 / 10 = 3.5
+		},
+		"(3 pairs of 3 records, three asset pool) interpolate: in between second and third record": {
+			recordsToSet: []types.TwapRecord{
+				tapRecordAB, tapRecordAC, tapRecordBC,
+				tPlus10sp5TapRecordAB, tPlus10sp5TapRecordAC, tPlus10sp5TapRecordBC,
+				tPlus20sp2TapRecordAB, tPlus20sp2TapRecordAC, tPlus20sp2TapRecordBC},
+			ctxTime: tPlusOneMin,
+			input:   makeSimpleTapTwapInput(baseTime.Add(15*time.Second), baseTime.Add(25*time.Second), quoteAssetA),
+			expTwap: []sdk.Dec{sdk.NewDecWithPrec(35, 1), sdk.NewDecWithPrec(35, 1), sdk.NewDecWithPrec(35, 1)}, // 5 for 5s, 2 for 5 = 35 / 10 = 3.5
 		},
 		// interpolate in time closer to second record
-		"(3 record) interpolate: get twap closer to second record": {
+		"(3 records) interpolate: get twap closer to second record": {
 			recordsToSet: []types.TwapRecord{baseRecord, tPlus10sp5Record, tPlus20sp2Record},
 			ctxTime:      tPlusOneMin,
 			input:        makeSimpleTwapInput(baseTime.Add(15*time.Second), baseTime.Add(30*time.Second), quoteAssetA),
-			expTwap:      sdk.NewDec(3), // 5 for 5s, 2 for 10s = 45 / 15 = 3
+			expTwap:      []sdk.Dec{sdk.NewDec(3)}, // 5 for 5s, 2 for 10s = 45 / 15 = 3
+		},
+		"(3 pairs of 3 records, three asset pool) interpolate: get twap closer to second record": {
+			recordsToSet: []types.TwapRecord{
+				tapRecordAB, tapRecordAC, tapRecordBC,
+				tPlus10sp5TapRecordAB, tPlus10sp5TapRecordAC, tPlus10sp5TapRecordBC,
+				tPlus20sp2TapRecordAB, tPlus20sp2TapRecordAC, tPlus20sp2TapRecordBC},
+			ctxTime: tPlusOneMin,
+			input:   makeSimpleTapTwapInput(baseTime.Add(15*time.Second), baseTime.Add(30*time.Second), quoteAssetA),
+			expTwap: []sdk.Dec{sdk.NewDec(3), sdk.NewDec(3), sdk.NewDec(3)}, // 5 for 5s, 2 for 10s = 45 / 15 = 3
```

Comment:
Asking to learn: I noticed that all the test cases for three asset pool have the same values for expTwap, which I don;t understand why in the first glance. Could you explain why we have the same expTwap for the testcases here?

Reply:
So AFAIU, the reason why we have the same results across all three TwapRecords is because we are making the same modifications to pair AB as we are for AB and BC. If we modified the spot price of one of the three pairs differently than the other two, than that spot price would have a differing result. I kept all three records with the same modifications for each record in order to retain my sanity. If we want to test more unique cases (which we definitely should!), we should do so after agreeing that this is/is not the right approach we want for the base TAP test structure.